sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java

```diff
@@ -148,18 +153,30 @@ public ShardingUpsertExecutor(ClusterService clusterService,
     private CompletionStage<BitSet> consumeIterator(BatchIterator batchIterator) {
         Row row = RowBridging.toRow(batchIterator.rowData());
         try {
-            while (batchIterator.moveNext()) {
-                onRow(row);
-                if (idxWithinBatch == batchSize) {
-                    CompletableFuture<BitSet> executeBatchFuture = execute(false);
+            while (collectionEnabled && batchIterator.moveNext()) {
+                if (idxWithinBatch == bulkSize) {
                     idxWithinBatch = 0;
-                    return executeBatchFuture.thenCompose(r -> consumeIterator(batchIterator));
+                    batchExecutionFutures.add(execute(false));
                 }
+                onRow(row);
+            }
+
+            if (collectionEnabled == false) {
+                // resume iterator consumption when the in-progress batches are done
```

Comment:
Execution is not disabled. The in progress batches continue to execute. We stop just the collection and resume it when the current in progress batches are done